docs/02-guides/customize-wysiwyg-platform.mdx

```diff
@@ -0,0 +1,200 @@
+  selector. See
+  [WYSIWYG dynamic styles](/docs/3.x/guides/display-wysiwyg-content#dynamic-styles)
+  for details.
+
+### Add a custom Magento Widget
+
+Custom widgets will be automatically parsed. However, you will still need to map
+the widget's `type` to custom React Components. If you don't, they will be
+ignored and nothing will be rendered in the final page.
+
+To do so, please override
+`theme/modules/Wysiwyg/MagentoWysiwyg/Widget/getWidgetComponent.js` in your own
+theme.
+
+Now, within your newly created `getWidgetComponent.js` file, you will be able to
+add your own behavior. For instance, if you've created a widget in Magento that
+uses the type `acme/product-preview`, you will need to update the `Widget.js`
+file likewise:
+
+```js title="theme/modules/Wysiwyg/MagentoWysiwyg/Widget/getWidgetComponent.js"
+import React from "react";
+import Spinner from "theme/components/atoms/Spinner";
+import loadable from "@loadable/component";
+
+const defaultWidgetsMap = {
+  "acme/product-preview": loadable(
+    () => import("./path/to/ProductPreview.js"),
+    {
+      fallback: <Spinner />,
+    }
+  ),
+};
+
+// ... rest of the file is intact
+```
+
+If you restart your application, you will notice that the new widget component
+is displayed. Hurray!
+
+However, in most cases, you will need to fetch data to display all the needed
+information in your widget. For instance, if the widget is
+`{% raw %}{{ widget type="acme/product-preview" sku="VSK12" }}{% endraw %}` you
+will want to fetch the product associated with the given SKU.
+
+#### Register your widget type at the GraphQL level
+
+1.  First make sure that your dependencies are up to date in your GraphQL module
+    ```ts title="extensions/acme-extension/modules/wysiwyg/index.ts"
+    export default createGraphQLModule({
+    namespace: "Magento2/Cms",
+    // highlight-start
+    dependencies: [
+      "Magento2/Wysiwyg", // ensure that Widget related features are available
+      "Magento2/Catalog/Product" // ensure that you can fetch a product in your Wysiwyg data
+    ],
+    // highlight-end
+    ```
+1.  Then you need to register a new WidgetData type in your schema
+    ```ts title="extensions/acme-extension/modules/wysiwyg/index.ts"
+    typeDefs: /* GraphQL */ `
+      type WidgetProductPreviewData implements WidgetData {
+        dataId: ID
+        product: Product
+      }
+    `,
+    ```
+1.  After that you can setup the resolvers to tell GraphQL how to fetch the
+    `product` field
+    ```ts title="extensions/acme-extension/modules/wysiwyg/resolvers.ts"
+    export default {
+      WidgetProductPreviewData: {
+        product: ({ node }, _, { loaders }) => {
+          const productSkuAttribute = node.attrs.find(
+            ({ name }) => name === "sku"
+          );
+          if (!productSkuAttribute) {
+            return null;
+          }
+          return loaders.Product.load(productSkuAttribute.value);
+        },
+      },
+    };
```

Comment:
I agree for typeDefs, but that is not true for resolvers. In fact, we still have `resolvers.js` files for most, if not all of our graphql modules.

I believe inlining resolvers will end up bloating most of `runtime.ts` files.

Reply:
Yes although we still have them, we will include it in the ADR that best practice is to inline them, this is so that they can benefit from the types of the `createGraphQLRuntime()` which eventually will include the context param types for all loaders, including config, user, services, etc..

Reply:
Oh I see. I'll update it then, but I believe the bloat point stands still. I think it'd be better to have a typed `resolvers.ts`.

Reply:
No because it needs to infer the types for the `contextEnhancer`, having them in separate files it won't be achievable.

Reply:
It should. That's an implementation detail. TS will infer it properly, no matter if it's from an import or not.

```ts
createGraphQLRuntime({ resolvers: { Query: { foo: () => "bar" }});
// is similar to
import resolvers from "./resolvers";
createGraphQLRuntime({ resolvers });
```

The reason why we recommend inline by default is because it's usually simpler to have everything in the same file, to follow the context of initialization and usage.

Because having small resolvers is a good practice, we saw that 80%+ of the time, both `resolvers` and `contextEnhancer ` fit in < 100 LOC. It's simpler to prevent an indirection in such cases.

But for big modules, keeping `resolvers` in a different file can make sense (even though in this case, it could be a smell of a module with too many responsibilities :angel:)

Reply:
That makes even more sense now, yes. Thanks for the clarification.